File: src/main/java/test/cafe/service/OrderService.java
package test.cafe.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import test.cafe.dto.OrderDto;
import test.cafe.dto.OrderItemDto;

import java.util.Optional;


/**
 *
 */
public interface OrderService {

    /**
     * Создание заказа.
     * @param orderDto DTO с данными для создания нового заказа
     * @return HTTP-ответ с DTO созданного заказа
     */
    OrderDto createOrder(OrderDto orderDto);

    /**
     * Изменение заказа.
     * @param id Идентификатор заказа.
     * @param orderDto DTO с данными изменяемого заказа.
     * @return HTTP-ответ с DTO измененного заказа.
     */
    Optional<OrderDto> editOder(Integer id, OrderDto orderDto);

    /**
     * Подтверждение заказа.
     * @param id Идентификатор заказа.
     * @return HTTP-ответ о подтверждении заказа.
     */
    Optional<OrderDto> confirmOrder(Integer id);

    /**
     * Отмена заказа.
     * @param id Идентификатор заказа.
     * @return HTTP-ответ об отмене заказа.
     */
    Optional<OrderDto> cancelOrder(Integer id);

    /**
     * Добавление позиции заказа.
     * @param orderId Идентификатор заказа.
     * @param orderItemDto DTO с данными позиций заказа.
     * @return HTTP-ответ с DTO добавленной позицией заказа.
     */
    Optional<OrderItemDto> addItem(Integer orderId, OrderItemDto orderItemDto);

    /**
     * Удаление позиции заказа.
     * @param orderId Идентификатор заказа.
     * @param itemId Идентификатор позиции заказа.
     * @return Пустой элемент.
     */
   boolean deleteItem(Integer orderId, Integer itemId);

    /**
     * Изменение позиции заказа.
     * @param orderId Идентификатор заказа.
     * @param itemId Идентификатор позиции заказа.
     * @param orderItemDto DTO с данными позиций заказа.
     * @return HTTP-ответ с DTO измененной позицией заказа.
     */
    Optional<OrderItemDto> editItem(Integer orderId, Integer itemId, OrderItemDto orderItemDto);

    /**
     * Получение списка заказов.
     * @param pageable Страница списка заказов.
     * @return Страница заказов.
     */
    Page<OrderDto> listOrders(Pageable pageable);

    /**
     * Получение заказа.
     * @return Optional-объект DTO заказа.
     */
    Optional<OrderDto> getOrder(Integer id);

    /**
     * Получение позиций заказа по идентификатору заказа
     *
     * @param orderId Идентификатор заказа.
     * @param pageable Страница списка позиций заказов.
     * @return Страница позиций заказа.
     */
    Page<OrderItemDto> listOrderItems(Integer orderId, Pageable pageable);

    /**
     * Получение позиции заказа.
     *
     * @param orderId идентификатор заказа
     * @param itemId  идентификатор позиции заказа
     * @return Optional-объект DTO позиции заказа
     */
    Optional<OrderItemDto> getItem(Integer orderId, Integer itemId);
}

File: coffee-frontend/src/app/service/order.service.ts
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {Order} from "../model/order.model";
import {ContentPage} from "../model/content-page.model";
import {OrderItem} from "../model/order-item.model";
import {
  CANCEL_URL_PART,
  CONFIRM_URL_PART,
  CREATE_URL_PART,
  EDIT_URL_PART,
  GET_URL_PART,
  ITEMS_URL_PART,
  LIST_URL_PART,
  ORDERS_LIST_URL,
  ORDERS_URL
} from "../constant/api-constants";
import {buildPageParams} from "../util/http-utlis";
import {Injectable} from "@angular/core";

@Injectable({
  providedIn: 'root'
})
export class OrderService {

  constructor(private httpClient: HttpClient) {
  }

  /**
   * Получение страницы заказов.
   *
   * @param page - номер страницы
   * @param size - размер страницы
   */
  public getOrders(page: number, size: number): Observable<ContentPage<Order>> | null {
    const url = ORDERS_LIST_URL
    const params = buildPageParams(page, size);
    return this.httpClient.post<ContentPage<Order>>(url, null, {params: params});
  }

  /**
   * Получение заказа.
   *
   * @param id - идентификатор заказа.
   */
  public getOrder(id: number): Observable<Order> | null {
    const url = `${ORDERS_URL}/${id}${GET_URL_PART}`; // Интерполяция
    return this.httpClient.get<Order>(url);
  }

  /**
   * Получение позиции заказа
   *
   * @param orderId - идентификатор заказа.
   * @param itemId - идентификатор позиции заказа
   */
  public getOrderItem(orderId: number, itemId: number): Observable<OrderItem> | null {
    const url = `${ORDERS_URL}/${orderId}${ITEMS_URL_PART}/${itemId}`;
    return this.httpClient.get<OrderItem>(url);
  }

  /**
   * Создание заказа.
   *
   * @param order - данные для создания нового заказа.
   */
  public createOrder(order: Order): Observable<Order> | null {
    const url = `${ORDERS_URL}${CREATE_URL_PART}`;
    return this.httpClient.post<Order>(url, order);
  }

  /**
   * Изменение заказа.
   *
   * @param id - идентификатор заказа.
   * @param order - данные изменяемого заказа.
   */
  public editOrder(id: number, order: Order): Observable<Order> | null {
    const url = `${ORDERS_URL}/${id}${EDIT_URL_PART}`;
    return this.httpClient.put<Order>(url, order);
  }

  /**
   * Подтверждение заказа.
   *
   * @param id - идентификатор заказа.
   */
  public confirmOrder(id: number): Observable<Order> | null {
    const url = `${ORDERS_URL}/${id}${CONFIRM_URL_PART}`;
    return this.httpClient.put<Order>(url, null);
  }

  /**
   * Отмена заказа.
   *
   * @param id - идентификатор заказа.
   */
  public cancelOrder(id: number): Observable<Order> | null {
    const url = `${ORDERS_URL}/${id}${CANCEL_URL_PART}`;
    return this.httpClient.put<Order>(url, null);
  }

  /**
   * Добавление позиции заказа.
   *
   * @param orderId - идентификатор заказа.
   * @param orderItem - данные позиций заказа.
   */
  public addItem(orderId: number, orderItem: OrderItem): Observable<OrderItem> | null {
    const url = `${ORDERS_URL}/${orderId}${ITEMS_URL_PART}`;
    return this.httpClient.post<OrderItem>(url, orderItem);
  }

  /**
   * Уаление позиции заказа.
   *
   * @param orderId - идентификатор заказа.
   * @param itemId - идентификатор позиции заказа
   */
  public deleteItem(orderId: number, itemId: number): Observable<void> | null {
    const url = `${ORDERS_URL}/${orderId}${ITEMS_URL_PART}/${itemId}`;
    return this.httpClient.delete<void>(url);
  }

  /**
   * Изменеие позиции заказа.
   *
   * @param orderId - идентификатор заказа.
   * @param itemId - идентификатор позиции заказа
   * @param orderItem - данные позиций заказа.
   */
  public editItem(orderId: number, itemId: number, orderItem: OrderItem): Observable<OrderItem> | null {
    const url = `${ORDERS_URL}/${orderId}${ITEMS_URL_PART}/${itemId}`;
    return this.httpClient.put<OrderItem>(url, orderItem);
  }

  /**
   * Получение страницы позиций заказа
   *
   * @param id - идентификатор заказа
   * @param page - номер страницы
   * @param size - размер страницы
   */
  public getOrderItems(id: number, page: number, size: number): Observable<ContentPage<OrderItem>> | null {
    const url = `${ORDERS_URL}/${id}${ITEMS_URL_PART}${LIST_URL_PART}`;
    const params = buildPageParams(page, size);
    return this.httpClient.post<ContentPage<OrderItem>>(url, null, {params: params});
  }
}
